include/nbl/video/IPhysicalDevice.h

```diff
@@ -755,4 +766,28 @@ namespace std
     };
 }
 
+namespace nbl::video
+{
+    inline uint64_t IPhysicalDevice::SBufferFormatPromotionRequestHash::operator()(const SBufferFormatPromotionRequest& r) const {
+        uint64_t msb = uint64_t(std::hash<IPhysicalDevice::SFormatBufferUsages::SUsage>()(r.usages));
+        return (msb << 32u) | r.originalFormat;
+    }
+
+    inline uint64_t IPhysicalDevice::SImageFormatPromotionRequestHash::operator()(const SImageFormatPromotionRequest& r) const {
+        uint64_t msb = uint64_t(std::hash<IPhysicalDevice::SFormatImageUsages::SUsage>()(r.usages));
+        return (msb << 32u) | r.originalFormat;
+    }
+
+    inline bool IPhysicalDevice::SBufferFormatPromotionRequestEqualTo::operator()(const SBufferFormatPromotionRequest& l, const SBufferFormatPromotionRequest& r) const
+    {
+        return l.originalFormat == r.originalFormat && l.usages == r.usages;
+    }
+
+     inline bool IPhysicalDevice::SImageFormatPromotionRequestEqualTo::operator()(const SImageFormatPromotionRequest& l, const SImageFormatPromotionRequest& r) const
+    {
+        return l.originalFormat == r.originalFormat && l.usages == r.usages;
+    }
```

Comment:
is there any problems with having these inline where the functions are declared?

Reply:
std::hash is declared above. its why I've declared it after that.

Reply:
ok I guess that solves some issues :D

Reply:
how does it work in a template though :?

Reply:
what you did is correct otherwise we'd have to "foward" declare the std::hash template which would have been hard

Reply:
just curious how you declare something ahead of std::hash in a template but if its written out then the compiler will complain about it